README.md

|        | Status           |
| ------------- |:-------------:|
| pypi      | [![openwrt_luci_rpc](https://img.shields.io/pypi/v/openwrt_luci_rpc.svg)](https://travis-ci.org/fbradyirl/openwrt-luci-rpc/) |
| travis    | [![Build Status](https://img.shields.io/travis/fbradyirl/openwrt-luci-rpc.svg)](https://travis-ci.org/fbradyirl/openwrt-luci-rpc/)      |
| docs |  [![Build Status](https://readthedocs.org/projects/openwrt-luci-rpc/badge/?version=latest)](https://openwrt-luci-rpc.readthedocs.io/en/latest/?badge=latest)    |


Python3 module for interacting with the OpenWRT Luci RPC interface


* Free software: Apache Software License 2.0
* Documentation: https://openwrt-luci-rpc.readthedocs.io.


Features
--------

* Allows you to use the Luci RPC interface to fetch connected devices on your OpenWrt based router.
* Supports 17.X & 18.X or newer releases of OpenWrt.


Usage
--------

You will first need to install the package `luci-mod-rpc` on your OpenWrt router.

```
opkg update
opkg install luci-mod-rpc
```

Once that is done, you can use this module to interact with the RPC interface.


### Install

```bash
pip install openwrt-luci-rpc
```

### Use

```python
from openwrt_luci_rpc import OpenWrtRpc

router = OpenWrtRpc('http://192.168.1.1', 'root', 'mypassword')

```

### Development

See [contributing guide](CONTRIBUTING.rst).


HISTORY.rst
=======
History
=======

0.1.0 (2019-02-19)
------------------

* Initial release. Not working. Do not use.

0.2.0 (2019-02-19)
------------------

* Initial release. Not working. Do not use.

0.3.0 (2019-02-19)
------------------

* First working release on PyPI.

0.4.8 (2019-02-20)
------------------

* Fix travis deploy repo, once more.

0.4.9 (2019-02-20)
------------------

* Fix README formatting for docs.

0.4.10 (2019-02-20)
-------------------

* Fix links to github repo

0.4.11 (2019-02-20)
-------------------

* Updated usage in README.

0.4.12 (2019-02-20)
-------------------

* Fix typo in README.

0.4.13 (2019-02-20)
-------------------

* Fix README parse issue.

1.0.0 (2019-02-21)
------------------

* Updated to return all device information.
* Added extra unit tests.

1.0.1 (2019-02-22)
------------------

* Minor bug fixes and extra unit tests

1.0.2 (2019-02-22)
------------------

* Logging fixes around token refresh

1.0.3 (2019-02-22)
------------------

* Fix History.rst formatting issue.

1.0.4 (2019-02-22)
------------------

* Take SSL param in init now
* Fix some typos

1.0.5 (2019-02-24)
------------------

* As a convenience, as the router IP as the host
  for every device. Can be useful when a network has more
  than one router.

1.0.6 (2019-02-26)
------------------

* Only a docs update.

1.0.7 (2019-02-26)
------------------

* Only a docs update.


1.1.0 (2019-07-16)
------------------

* Compatibility with 15.05.


1.1.1 (2019-09-19)
------------------

* Adding packaging dependancy.

1.1.2 (2019-10-11)
------------------

* Add option to enable or disable HTTPS verification.
* Suppress "InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised." warnings.
* Reduce number of info log messages.
* Minor code clean-up to address flake8 warnings.


1.1.3 (2020-04-20)
------------------
* Add packaging


1.1.5 (2020-09-01)
------------------
* Unpin pip versions. Allow higher. Fixes #42 https://github.com/fbradyirl/openwrt-luci-rpc/issues/42


1.1.6 (2020-09-01)
------------------
* Fixes home-assistant/core/issues/38870, home-assistant/core/issues/39498
* Fixes fbradyirl/openwrt-luci-rpc/issues/33

1.1.7 (2020-12-08)
------------------
* Switch pipeline to Github actions

1.1.8 (2021-03-12)
------------------
* Support newer openwrt snapshot versions (#45)
* Fix tox in github actions and also test py38 (#46)

1.1.9 (2021-03-21)
------------------
* Publish to pypi only once in github actions (#47)

1.1.10 (2021-03-21)
-------------------
* Also trigger the Github Actions on push of tags (#49)

1.1.11 (2021-03-21)
-------------------
* Add rst-linter pre-commit hook to prevent committing errors to HISTORY.rst #50


1.1.12 (2022-09-11)
-------------------
* Includes latest fixes to support newer versions of OpenWrt


1.1.16 (2023-03-09)
-------------------
* Downgrade logging level for version detection (#57)
* Modify test


test.py

import logging
import sys, os

root = logging.getLogger()
root.setLevel(logging.DEBUG)
log = logging.getLogger(__name__)

handler = logging.StreamHandler(sys.stdout)
handler.setLevel(logging.DEBUG)
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
handler.setFormatter(formatter)
root.addHandler(handler)

from openwrt_luci_rpc import OpenWrtRpc

env_name_ip_address = "OPENWRT_IP"
env_name_password = "OPENWRT_PASSWORD"

if not os.getenv(env_name_ip_address):
    log.error("You need to set the router ip environment variable. e.g. ")
    log.error("export {}=<YOUR_ROUTER_IP>".format(env_name_ip_address))
    exit(1)


if not os.getenv(env_name_password):
    log.error("You need to set the router password environment variable. e.g. ")
    log.error("export {}=<YOUR_ROUTER_PASS>".format(env_name_password))
    exit(1)


print('test valid')
r = OpenWrtRpc(os.getenv(env_name_ip_address), 'root', os.getenv(env_name_password), False)

# print('test invalid')
# r = OpenWrtRpc()


r.get_all_connected_devices(only_reachable=False)

#import ipdb
#ipdb.set_trace()
#
# r.get_all_connected_devices(only_reachable=False)
#
# import ipdb
# ipdb.set_trace()
#
# r.get_all_connected_devices(only_reachable=False)
#
# import ipdb
# ipdb.set_trace()


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for openwrt_luci_rpc."""


tests/test_version.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `openwrt_luci_rpc` package."""


import unittest
from openwrt_luci_rpc.utilities import is_legacy_version
from packaging import version


class TestOpenwrtLuciRPC(unittest.TestCase):
    """Tests for `openwrt_luci_rpc` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_is_legacy_version(self):
        """Test comparing versions works as expected."""

        assert is_legacy_version(version.parse("15.05")) is True
        assert is_legacy_version(version.parse("15.05.1")) is True
        assert is_legacy_version(version.parse("17.01")) is True
        assert is_legacy_version(version.parse("17.01.6")) is True

        assert is_legacy_version(version.parse("18.06")) is False
        assert is_legacy_version(version.parse("18.06.9")) is False
        assert is_legacy_version(version.parse("19.07.7")) is False


openwrt_luci_rpc/__init__.py
# -*- coding: utf-8 -*-

"""Top-level package for openwrt-luci-rpc."""

__author__ = """Finbarr Brady"""
__email__ = 'fbradyirl@github.io'
__version__ = '1.1.16'

from .openwrt_luci_rpc import OpenWrtLuciRPC
from .constants import Constants


class OpenWrtRpc:
    """
    Class to interact with OpenWrt router running luci-mod-rpc package.
    """

    def __init__(self, host_url=Constants.DEFAULT_LOCAL_HOST,
                 username=Constants.DEFAULT_USERNAME,
                 password=Constants.DEFAULT_PASSWORD,
                 is_https=Constants.DEFAULT_HTTPS,
                 verify_https=Constants.DEFAULT_VERIFY_HTTPS):
        """
        Initiate an instance with a default local ip (192.168.1.1)
        :param host_url: string - host url. Defaults to 192.168.1.1
        :param username: string - username. Defaults to root
        :param password: string - password. Default is blank
        :param is_https: boolean - use https? Default is false
        :param verify_https: boolean - verify https? Default is true
        """
        self.router = OpenWrtLuciRPC(host_url, username, password,
                                     is_https, verify_https)

    def is_logged_in(self):
        """Returns true if a token has been aquired"""
        return self.router.token is not None

    def get_all_connected_devices(self,
                                  only_reachable=Constants.DEFAULT_ONLY_REACH,
                                  wlan_interfaces=Constants.DEFAULT_WLAN_IF):
        """Get details of all devices"""
        return self.router.get_all_connected_devices(
            only_reachable=only_reachable, wlan_interfaces=wlan_interfaces)


openwrt_luci_rpc/constants.py
class Constants:

    DEFAULT_LOCAL_HOST = "192.168.1.1"
    DEFAULT_USERNAME = "root"
    DEFAULT_PASSWORD = ""
    DEFAULT_TIMEOUT = 30
    DEFAULT_HTTPS = False
    DEFAULT_VERIFY_HTTPS = True
    DEFAULT_ONLY_REACH = True
    DEFAULT_ARP = True
    DEFAULT_WIFI = False
    DEFAULT_WLAN_IF = 'wlan0'

    LUCI_RPC_LOGIN_PATH = '{}/cgi-bin/luci/rpc/auth'
    LUCI_RPC_SYS_PATH = '{}/cgi-bin/luci/rpc/sys'
    LUCI_RPC_UCI_PATH = '{}/cgi-bin/luci/rpc/uci'

    # Only available in 18.06+
    LUCI_RPC_IP_PATH = '{}/cgi-bin/luci/rpc/ip'

    # Make old and new keys consistent for 17 & 18 releases
    # Spaces and dots are removed so keep them out of here also.
    MODERN_KEYS = {
        "HW_address": "mac",
        "IP_address": "ip",
        "dest": "ip"
    }

    def __init__(self):
        pass


openwrt_luci_rpc/utilities.py
from .constants import Constants
from packaging import version
import logging

log = logging.getLogger(__name__)


def normalise_keys(result):
    """Replace 17.06 keys with newer ones."""
    if result is not None:

        # named tuple keys cannot have spaces or begin with dots. Replace
        # those with underscores
        result = \
            {k.replace(".", "_").replace(" ", "_"): v for k, v in result.items()}  # noqa: E501

        for old_key, new_key in Constants.MODERN_KEYS.items():
            if old_key in result:
                result[new_key] = result[old_key]
                del result[old_key]

    return result


def get_hostname_from_dhcp(dhcp_result, mac):
    """Determine the hostname for this mac."""
    if dhcp_result is not None and isinstance(dhcp_result, dict):
        host = [x for x in dhcp_result.values()
                if x['.type'] == 'host'
                and 'mac' in x
                and 'name' in x
                and x['mac'].upper() == mac]

        if host:
            log.debug("DNS name lookup for mac {} "
                      "found {}".format(mac, host[0]['name']))
            return host[0]['name']

    return None


def is_legacy_version(owrt_version):
    return (
        "snapshot" not in owrt_version.public and
        owrt_version < version.parse("18.06")
    )


openwrt_luci_rpc/openwrt_luci_rpc.py
# -*- coding: utf-8 -*-

"""
Support for OpenWrt (luci) routers.

For more details about this platform, please refer to the documentation at
https://home-assistant.io/components/device_tracker.luci/
"""

import requests
import json
import logging

from packaging import version
from collections import namedtuple
from requests.packages.urllib3.exceptions import InsecureRequestWarning
from openwrt_luci_rpc import utilities
from .constants import Constants
from .exceptions import InvalidLuciTokenError, \
    LuciRpcMethodNotFoundError, InvalidLuciLoginError, \
    LuciRpcUnknownError, PageNotFoundError, LuciConfigError

log = logging.getLogger(__name__)


class OpenWrtLuciRPC:

    def __init__(self, host, username, password, is_https, verify_https):
        """
        Initiate an API request with all parameters
        :param host: string
        :param username: string
        :param password: string
        :param is_https: boolean
        :param verify_https: boolean
        """

        if not host:
            raise LuciConfigError('host cannot be empty. '
                                  'Use the IP or hostname of your'
                                  'OpenWrt router')

        if not is_https:
            protocol = 'http'
        else:
            protocol = 'https'

            if not verify_https:
                requests.packages.urllib3.disable_warnings(
                    InsecureRequestWarning
                )

        self.host = host
        self.host_api_url = '{}://{}'.format(protocol, host)
        self.username = username
        self.password = password
        self.verify_https = verify_https
        self.session = requests.Session()
        self.token = None
        self.owrt_version = None
        self._refresh_token()
        self.is_legacy_version, self.arp_call \
            = self._determine_if_legacy_version()

    def _refresh_token(self):
        """Get authentication token for the given configuration."""
        auth_url = Constants.\
            LUCI_RPC_LOGIN_PATH.format(self.host_api_url)
        self.token = self._call_json_rpc(auth_url, 'login',
                                         self.username, self.password)
        log.info("Luci RPC login was successful")

    def _determine_if_legacy_version(self):
        """
        * Checks to see if we are running a pre-18.06 version

        :return: tuple, with True if legacy and the URL to
                use to lookup devices
        """

        log.info("Getting OpenWRT version")

        # NEW METHOD TO DETERMINE OPENWRT VERSION EXACTLY
        # TODO: Check as non-root user

        # get VERSION_ID from os-release if exists or get
        # DISTRIB_RELEASE from openwrt_release
        shell_command = "if [ -f \"/etc/os-release\" ]; \
                            then awk -F= '$1==\"VERSION_ID\" \
                            { print $2 ;}' \
                            /etc/os-release; \
                            else awk -F= '$1==\"DISTRIB_RELEASE\" \
                            { print $2 ;}' \
                            /etc/openwrt_release; fi | \
                            tr -d \\'\\\""

        rcp_sys_version_call = Constants.\
            LUCI_RPC_SYS_PATH.format(self.host_api_url), "exec"

        try:
            content = self._call_json_rpc(rcp_sys_version_call[0],
                                          rcp_sys_version_call[1],
                                          shell_command)   # type: str

            content = content.replace("\n", "")

            if content is None:
                raise LuciRpcUnknownError("could not \
                determine openwrt version")

            self.owrt_version = version.parse(content.strip())
        except InvalidLuciLoginError:
            log.info("Refreshing login token")
            self._refresh_token()
            return self._determine_if_legacy_version()
        except Exception:
            log.debug("Could not determine OpenWRT version, \
                         defaulting to version 18.06")
            self.owrt_version = version.parse("18.06")

        rpc_sys_arp_call = Constants.\
            LUCI_RPC_SYS_PATH.format(self.host_api_url), 'net.arptable'
        rpc_ip_call = Constants.\
            LUCI_RPC_IP_PATH.format(
                self.host_api_url), 'neighbors', {"family": 4}

        if utilities.is_legacy_version(self.owrt_version):
            return True, rpc_sys_arp_call
        else:
            return False, rpc_ip_call

    def get_all_connected_devices(self, only_reachable, wlan_interfaces):
        """
        Get details of all connected devices.

        Notes around newer OpenWrt releases (18.06+)

            Do not use `reachable` or `stale` values
            as they flap constantly even
            when the device is inside the network.
            The very existence of the mac in the results
            is enough to determine the "device is home"

        :param only_reachable: boolean, if true,
               only return devices which are reachable
               (this is for 17.06 or earlier only. 18+
               does not have a proper ability to determine
               this, as above)
        """
        log.debug("Checking for connected devices")
        last_results = []
        # rpc_sys__winfo_call = Constants.\
        #     LUCI_RPC_SYS_PATH.format(self.host_api_url), \
        #                       'wifi.getiwinfo', wlan_interfaces
        rpc_uci_call = Constants.LUCI_RPC_UCI_PATH.format(
            self.host_api_url), 'get_all', 'dhcp'

        try:
            # First, try find the associated wifi devices
            # winfo_result = self._call_json_rpc(*rpc_sys__winfo_call)
            arp_result = self._call_json_rpc(*self.arp_call)
            dhcp_result = self._call_json_rpc(*rpc_uci_call)
        except InvalidLuciTokenError:
            log.info("Refreshing login token")
            self._refresh_token()
            return self.get_all_connected_devices(only_reachable,
                                                  wlan_interfaces)

        for device_entry in arp_result:
            if device_entry is None:
                continue

            device_entry = utilities.normalise_keys(device_entry)

            if "mac" not in device_entry:
                continue

            device_entry['hostname'] = utilities.get_hostname_from_dhcp(
                dhcp_result, device_entry['mac'])

            # As a convenience, add the router IP as the host
            # for every device. Can be useful when a network has more
            # than one router.
            if "host" not in device_entry:
                device_entry['host'] = self.host

            device = namedtuple("Device", device_entry.keys())(
                *device_entry.values())

            if self.is_legacy_version:
                if "Flags" in device_entry and only_reachable:
                    # Check if the Flags for each device contain
                    # NUD_REACHABLE and if not, skip.
                    if not int(device_entry['Flags'], 16) & 0x2:
                        continue
            else:
                if "reachable" in device_entry and only_reachable:
                    if device_entry['reachable'] is False:
                        continue

            last_results.append(device)

        log.debug(last_results)
        return last_results

    def _call_json_rpc(self, url, method, *args, **kwargs):
        """Perform one JSON RPC operation."""
        data = json.dumps({'method': method, 'params': args})

        # pass token to make it work with versions < 17
        if self.token is not None:
            url += "?auth=" + self.token

        log.debug("_call_json_rpc : %s" % url)
        res = self.session.post(url,
                                data=data,
                                timeout=Constants.DEFAULT_TIMEOUT,
                                verify=self.verify_https,
                                **kwargs)

        if res.status_code == 200:
            result = res.json()
            try:
                content = result['result']

                if content is not None:
                    return content

                elif result['error'] is not None:
                    # On 18.06, we want to check for error 'Method not Found'
                    error_message = result['error']['message']
                    error_code = result['error']['code']
                    if error_code == -32601:
                        raise LuciRpcMethodNotFoundError(
                            "method: '%s' returned an "
                            "error '%s' (code: '%s).",
                            method, error_message, error_code)
                else:
                    log.debug("method: '%s' returned : %s" % (method, result))
                    # Authentication error
                    raise InvalidLuciLoginError("Failed to authenticate "
                                                "with Luci RPC, check your "
                                                "username and password.")

            except KeyError:
                raise LuciRpcUnknownError("No result in response from luci")

        elif res.status_code == 401:
            raise InvalidLuciLoginError("Failed to authenticate "
                                        "with Luci RPC, check your "
                                        "username and password.")
        elif res.status_code == 403:
            raise InvalidLuciTokenError("Luci responded "
                                        "with a 403 Invalid token")
        elif res.status_code == 404:
            raise PageNotFoundError("404 returned "
                                    "from %s. Ensure you have "
                                    "installed package "
                                    "`luci-mod-rpc`." % url)

        raise LuciRpcUnknownError("Invalid response from luci: %s", res)
